### PR TITLE
[ports/stm32]: Added generic MP IRQ callback and example for UART (Version II).

### PR DESCRIPTION
**Pull request by tobbad**
# Rework of PR #3854
This is a rework of PR #3854 (which is kept for reference) which is more in sync with the existing solution for CC3200 (thanks to Daniel Campora). For general remarks see the old PR. I think this solution is much better than my PR #3854 . All credits go to Daniel. I kept the code for the STM port in a separate subfolder (`misc`) as he does.

# Aim of PR
At several locations in the stm32/port (maybe in other ports too) there is a call from the interrupt context to a user defined micropython routine/method. As a result we have some duplicated code (see Handle_EXTI_Irq, timer_handle_irq_channel) and some ToDo (se eg. pin_irq method). Further the way in timer the callback is set-up should be unified to the way it is done elsewhere.

There are some discussions in the forum on this topic(Event/IRQ callback):
https://forum.micropython.org/viewtopic.php?f=2&t=1448
https://forum.micropython.org/viewtopic.php?f=6&t=4414
https://forum.micropython.org/viewtopic.php?f=6&t=1312
https://forum.micropython.org/viewtopic.php?f=2&t=194

Issues #1642 and the discussion in the Hardware API (https://github.com/micropython/micropython/wiki/Hardware-API) on IRQ

My proposed MP irq module tries to keep close to the API of pin.irq.

# Improvements to PR 3854
IRQs are stored in a global list (modifications in `mpconfigport.h`) which is initialized in main. In contrast to Daniels solution, I do not reallocate an IRQ if a new handler or trigger is registered on it. I allocate a `mp_irq_obj_t` and keep it in the global list forever (or till the next reboot) to avoid memory fragmentation. 

Further in contrast to Daniels solution I do not support the `priority` and the `wake` parameter on the irq function call. However I do introduce a `hard` parameter which allows to decide if the the call to a micropython function (`hard=false`) is scheduled in the C code in the IRQ or a if the call to a micropython callback is executed in the IRQ context with all its constraints (No allocation ...). Schedule a micropyton call in C only takes 2us which is much faster than schedule a call from uPy (of course...).

Further I added a `trigger()` function on irq to obtain the currently active IRQs and therefore I renamed the `flags` callback to the IRQ generating module to `info` and added a parameter to get a certain property from the IRQ-generating module. 

An IRQ has to be explicitly enabled. Adding a handler and a trigger as parameter to the irq call does not enable it. I think this allows a better control *when* you'd like to start processing IRQ.

In the uart MP module I had to add a `mp_irq_map` as enable the IRQ on L4 is distributed over several registers and on F4 it is uint32_t wide. So in my opinion this was the most obvious way. 

In my opinion it would be possible to unit the two solutions (CC3200, STM) into a single one where there is no code duplication (duplication is not good). The `mp_irq_priorities` in Daniels could be introduced by injection, we have to settle on a reallocation schema for the IRQ struct (`mp_irq_obj_t`) and if we need a separate file for error strings.

# Example usage of IDLE IRQ for UART
On STM32 the UART RX IDLE IRQ is fired if after receiving a stream of input data there is for the duration of 1 Byte no new data received (RX signal is idle). This feature can be used to answer fast to a command received from a communication partner and therefore avoid  the other side to run into a TIMEOUT.

Further (artificial) MicroPython IRQ could be added: Eg Testing the RX or TX buffer on the filling level to fire a call to the callback for Buffer nearly empty or full. This feature is not implemented however. Fire the MP callback on the HW generated TX Empty is not recommend as such high rate IRQ would stall MicroPython.

Add following class:
```
import micropython

class UART_IDLE:
    
    USE_HARD = False
    
    def __init__(self, uart):
        self._uart = uart
        self._uart.write("Hi there\n\r")
        self._buffer = bytearray(64)
        self._answer = b"Ready \r\n"
        self._normal_context_ref = self.normal_context
        
    def set_up_cb(self):
        flags = self._uart.IRQ_RX_IDLE
        if UART_IDLE.USE_HARD:
            self._uart.irq(self.irq_idle_callback, flags, UART_IDLE.USE_HARD)
        else:
            self._uart.irq(self.normal_context, self._uart.IRQ_RX_IDLE, UART_IDLE.USE_HARD)
        
    def irq_idle_callback(self, other):
        micropython.schedule(self._normal_context_ref, 0)
    
    def normal_context(self, value):
        cnt=self._uart.readinto(self._buffer)
        self._uart.write(self._answer)
        data = "".join([chr(i) if 32<=i<128 else "."  for i in self._buffer[:cnt] ])
        print("Get UART data in normal context: %s" % data)
        self._uart.write("Received: \"%s\"\r\n" % data)
```
... and use following code:
```
uart = pyb.UART(2, baudrate = 115200, timeout=1)
dut=UART_IDLE(uart)
dut.set_up_cb()
callback= uart.irq()
callback()    # call the callback
print("Trigger flag %d " % uart.irq().trigger()) # Print active set triggers
# Enable IRQ
uart.irq().enable()
```
I am looking for feedback for this improved solution.

### CODE REVIEW

**Comment by tobbad** [comment]
Any Feedback/News on this?

**Comment by peterhinch** [comment]
I wonder if you could clarify the intent here. If a UART is receiving data intermittently, when does the RX idle callback run? Does it occur when the UART timeout expires?

At the Python level I would suggest that the most useful callbacks are TX buffer empty, RX timeout and RX buffer "nearly" full, where "nearly" is user definable.

A tx buffer empty callback would be useful in cases where the dataset to be sent is larger than the buffer: the callback would feed the buffer.

**Comment by tobbad** [comment]
@peterhinch: I hope I answered your questions. The TX IDLE callback could be realized by checking the empty space of the MP circular buffer. RX nearly full could be done in a similar way. 

The RX Idle implemented as example will not protect you from overwriting already received data if the received stream of data is larger than the UART circular buffer size.

**Comment by dpgeorge** [comment]
Sorry for the long delay responding here, and thanks for submitting a revised version.  In general it looks good and I think we can move forward with it after addressing a few points.

> My proposed MP irq module tries to keep close to the API of pin.irq.

Great!  Note that stm32, esp8266 and esp32 are pretty close in how their pin.irq() method works, and it would be good if they didn't have to change.  They have `handler`, `trigger`, and `hard` arguments, and the irq object has `__call__()` and `trigger()` methods (but not in stm32, it doesn't have an irq object yet).



> IRQs are stored in a global list (modifications in `mpconfigport.h`) which is initialized in main. In contrast to Daniels solution, I do not reallocate an IRQ if a new handler or trigger is registered on it. I allocate a `mp_irq_obj_t` and keep it in the global list forever (or till the next reboot) to avoid memory fragmentation.

Good idea.

> Further in contrast to Daniels solution I do not support the `priority` and the `wake` parameter on the irq function call.

Good, they can be added later if needed.

> However I do introduce a `hard` parameter

Good, this maintains consistency with existing pin.irq().

> Further I added a `trigger()` function on irq to obtain the currently active IRQs

Good, also maintains consistency with pin.irq().  But note that pin.irq().trigger() allows you to set the trigger by passing in a value (and will then return the previous trigger settings).

> and therefore I renamed the `flags` callback to the IRQ generating module to `info` and added a parameter to get a certain property from the IRQ-generating module.

Yes that's neat.  It could go further and have just a single, combined IRQ method for disable/enable/get-flags/get-trigger.

> An IRQ has to be explicitly enabled. Adding a handler and a trigger as parameter to the irq call does not enable it. I think this allows a better control _when_ you'd like to start processing IRQ.

This goes against the existing pin.irq() behaviour, which enables it when the method is called.  I'd much prefer to keep the existing behaviour, at least to minimise changes and maintain compatibility with existing code.  Also I think it's simpler to enable the IRQ by default.  If you don't want it enabled right away then you can pass in 0 for the trigger, then set the trigger to the correct values using the irq.trigger(value) method later one.

In fact I think irq.enable() and irq.disable() are not needed at all: just use irq.trigger(0) to disable and irq.trigger(value) to enable.  Furthermore, irq.init(...) is also not really necessary, instead just use uart.irq(...) to completely reinitialise the IRQ.  Not having these 3 methods keeps the number of methods down to a minimum, and they can be added later if really needed (but taking things away that are already there is much harder).

> In my opinion it would be possible to unit the two solutions (CC3200, STM) into a single one where there is no code duplication (duplication is not good). The `mp_irq_priorities` in Daniels could be introduced by injection, we have to settle on a reallocation schema for the IRQ struct (`mp_irq_obj_t`) and if we need a separate file for error strings.

Yes, and I think we can start right now in this PR by putting the generic mpirq.[ch] code in a common place (eg lib/utils/) instead of the ports/stm32 directory.

----

So there are 2 main things here: 1) what the user facing API is for IRQs; 2) how that is implemented underneath.  I think the implementation here in this PR is pretty good, it just needs to be refined to get the API in line with what exists in pin.irq() for stm32, esp8266 and esp32.  I realise that the cc3200's IRQ API is what is being taken as reference for this PR, but I think we should adapt that one rather than the stm32/esp8266/esp32.

In summary I suggest the IRQ API be the following:
```python
irq = periph.irq() # get IRQ object
irq = periph.irq(handler, trigger, hard) # configure and return IRQ
irq.trigger() # get current trigger
irq.trigger(value) # set trigger and return previous trigger value
irq.flags() # get current flags
```
Note that the docs already specify the periph.irq(...) behaviour (see eg machine.Pin.irq), which doesn't need to change.  The main thing here is to specify the IRQ object.  The above should be compatible with existing stm32/esp8266/esp32, and the cc3200 can be adapted to support this while still having its old methods (namely init, enable, disable).

**Comment by tobbad** [comment]
Thanks for the feedback. In first step I rebased it and will then modify the source based on your input - I think at the end of this week.

**Comment by tobbad** [comment]
I reworked the code according to your inputs. Some remarks
- The proposal of the [Hardware-API](https://github.com/micropython/micropython/wiki/Hardware-API) should be adapted to the current "way-to-go": (Remove init/enable/disable functions on the irq object, add trigger function on it)
- On an IRQ object the call() (uart.irq()()) is still there (Contrary to your summary)
So the usage changed slightly:
```python
import micropython

class UART_IDLE:
    
    USE_HARD = False
    READY="Ready\r\n"
    
    def __init__(self, uart):
        self._uart = uart
        self._uart.write("Hi there\n\r")
        self._buffer = bytearray(64)
        self._answer_irq = b"Irq \r\n"
        self._answer = b"Normal \r\n"
        self._normal_context_ref = self.normal_context
        
    def set_up_cb(self):
        flags = self._uart.IRQ_RX_IDLE
        if UART_IDLE.USE_HARD:
            self._uart.irq(self.irq_idle_callback, flags, UART_IDLE.USE_HARD)
        else:
            self._uart.irq(self.normal_context, flags, UART_IDLE.USE_HARD)
        
    def irq_idle_callback(self, other):
        self._uart.write(self._answer_irq)
        micropython.schedule(self._normal_context_ref, 0)
    
    def normal_context(self, value):
        cnt=self._uart.readinto(self._buffer)
        self._uart.write(self._answer)
        data = "".join([chr(i) if 32<=i<128 else "."  for i in self._buffer[:cnt] ])
        print("Get UART data in normal context: %s" % data)
        self._uart.write("Received: \"%s\"\r\n" % data)
```
And the usage (no enable() anymore which is done as default :-) 
```python
from uart_idle import UART_IDLE 
uart = pyb.UART(1, baudrate = 115200, timeout=1)
dut=UART_IDLE(uart)
dut.set_up_cb()
callback= uart.irq()
callback()    # call the callback
print("Trigger flag 0x%02x " % uart.irq().trigger()) # Print active set triggers
```
So we have following API:
```python
# get IRQ (or callback) object if it was instantiated before on this periph; None otherwise:
irq = periph.irq() 
# Create/Get IRQ Object and initialize it:
irq = periph.irq(handler_function, trigger_value, hard=True) 
irq.trigger() # get current trigger
irq.trigger(value) # set trigger and return previous trigger value
irq.flags() # get current flags
irq() # trigger the callback
```

**Comment by dpgeorge** [comment]
Thanks @tobbad for updating.  I'll do a detailed review soon.  For now, don't worry about rebasing on/merging with the latest master because there were a lot of changes to the UART code.

**Comment by dpgeorge** [comment]
Ok, I have tested this code on a PYBv1.0 and it works well, works as described here.  I needed to updated it due to recent refactoring of the uart.c code into uart.c+machine_uart.c, but no real problem there.

I think this is a good step forward, the API is the same as already exists on the Pin object (for stm32, esp8266, esp32) except that it adds the `flags()` method which is necessary to distinguish amongst multiple IRQ sources, and is how the cc3200 works.

I'm happy to merge this after I do some minor additional cleaning due to the refactor mentioned above.
